src/signatures/secp256k1_ecdsa.rs

```diff
@@ -0,0 +1,210 @@
+
+#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Address([u8; Self::LENGTH]);
+
+impl Address {
+    pub const LENGTH: usize = 20;
+}
+
+impl AsRef<[u8; Address::LENGTH]> for Address {
+    fn as_ref(&self) -> &[u8; Address::LENGTH] {
+        &self.0
+    }
+}
+
+impl From<[u8; Address::LENGTH]> for Address {
+    fn from(bytes: [u8; Address::LENGTH]) -> Self {
+        Self(bytes)
+    }
+}
+
+impl From<Address> for [u8; Address::LENGTH] {
+    fn from(address: Address) -> Self {
+        address.0
+    }
+}
+
+impl Hash for Address {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.hash(state);
+    }
+}
+
+#[derive(ZeroizeOnDrop)]
+pub struct SecretKey(k256::ecdsa::SigningKey);
+
+impl SecretKey {
+    pub const LENGTH: usize = 32;
+
+    #[cfg(feature = "rand")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "rand")))]
+    pub fn generate() -> Self {
+        let mut rng = rand::rngs::OsRng;
+        Self(k256::ecdsa::SigningKey::random(&mut rng))
+    }
+
+    #[cfg(feature = "rand")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "rand")))]
+    pub fn generate_with<R: rand::CryptoRng + rand::RngCore>(rng: &mut R) -> Self {
+        Self(k256::ecdsa::SigningKey::random(rng))
+    }
+
+    pub fn public_key(&self) -> PublicKey {
+        PublicKey(k256::ecdsa::VerifyingKey::from(&self.0))
+    }
+
+    pub fn to_bytes(&self) -> [u8; SecretKey::LENGTH] {
+        self.0.to_bytes().into()
+    }
+
+    pub fn try_from_bytes(bytes: [u8; SecretKey::LENGTH]) -> crate::Result<Self> {
+        k256::ecdsa::SigningKey::from_bytes(&bytes.into())
+            .map_err(|_| crate::Error::ConvertError {
+                from: "bytes",
+                to: "secp256k1 ecdsa secret key",
+            })
+            .map(Self)
+    }
+
+    // Signature operation can fail for some reason.
+    // Not sure what to do. Panic? Retry?
+    pub fn try_sign(&self, msg: &[u8]) -> crate::Result<Signature> {
+        self.0
+            .sign_recoverable(msg)
+            .map_err(|_| crate::Error::SignatureError { alg: "secp256k1 ecdsa" })
+            .map(|(sig, rid)| Signature(sig, rid))
+    }
+
+    pub fn sign(&self, msg: &[u8]) -> Signature {
+        self.try_sign(msg).expect("secp256k1 ecdsa sign failed")
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct PublicKey(k256::ecdsa::VerifyingKey);
+
+impl PublicKey {
+    pub const LENGTH: usize = 33;
+
+    pub fn verify(&self, sig: &Signature, msg: &[u8]) -> bool {
+        use k256::ecdsa::signature::Verifier;
+        self.0.verify(msg, &sig.0).is_ok()
+    }
+
+    pub fn to_bytes(self) -> [u8; PublicKey::LENGTH] {
+        let encoded_point = self.0.to_encoded_point(true);
+        let slice = encoded_point.as_bytes();
+        let mut bytes = [0_u8; PublicKey::LENGTH];
+        bytes.copy_from_slice(slice);
+        bytes
+    }
+
+    pub fn try_from_bytes(bytes: &[u8; PublicKey::LENGTH]) -> crate::Result<Self> {
```

Comment:
`try_from_bytes(self.to_bytes())` would work, right? Because one uses encoded point and the other sec1 bytes. I guess they can be interchangeable but why are we not using the same method for both?

Reply:
Good point. Internally, it should be the same. Do you think we should enforce compressed form?

Reply:
Alright. Added checks to only support SEC1 compressed public keys (with tags 2 or 3), compact public keys (tag = 5) are forbidden.